warehouse/oidc/models/activestate.py

```diff
@@ -0,0 +1,204 @@
+) -> bool:
+    # We expect a string formatted as follows:
+    #  org:<orgName>:project:<projectName>
+    # where :OPTIONAL-STUFF is a concatenation of other job context
+    # metadata. We currently lack the ground context to verify that
+    # additional metadata, so we limit our verification to just the ORG/REPO
+    # component.
+
+    # Defensive: ActiveState should never give us an empty subject.
+    if not signed_claim:
+        raise InvalidPublisherError("Missing 'subject' claim")
+
+    components = signed_claim.split(":")
+
+    if len(components) < 4:
+        raise InvalidPublisherError("Invalid 'subject' claim. Wrong format")
+
+    matches = (
+        f"{components[0]}:{components[1]}:{components[2]}:{components[3]}"
+        == ground_truth
+    )
+    if not matches:
+        raise InvalidPublisherError("Invalid 'subject' claim")
+
+    return True
+
+
+class ActiveStatePublisherMixin:
+    """
+    Common functionality for both pending and concrete ActiveState OIDC publishers.
+    """
+
+    organization = mapped_column(String, nullable=False)
+    activestate_project_name = mapped_column(String, nullable=False)
+    actor = mapped_column(String, nullable=False)
+    # 'actor' (The ActiveState platform username) is obstained from the user
+    # while configuring the publisher We'll make an api call to ActiveState to
+    # get the 'actor_id'
+    actor_id = mapped_column(String, nullable=False)
+    ingredient = mapped_column(String, nullable=True)
+
+    __required_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        "sub": _check_sub,
+        "organization": oidccore.check_claim_binary(str.__eq__),
+        "project": oidccore.check_claim_binary(str.__eq__),
+        "actor_id": oidccore.check_claim_binary(str.__eq__),
+        # This is the name of the builder in the ActiveState Platform that
+        # publishes things to PyPI.
+        "builder": oidccore.check_claim_invariant("pypi-publisher"),
+    }
+
+    __optional_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        # Should we check that the pypi projectname matches this ingredient
+        # name?
+        "ingredient": oidccore.check_claim_binary(str.__eq__),
+    }
+
+    __unchecked_claims__ = {
+        "actor",
+        "artifact_id",
+        "organization_id",
+        "project_id",
+        "project_path",
+        "project_visibility",
+    }
+
+    @staticmethod
+    def __lookup_all__(klass, signed_claims: SignedClaims):
+        return Query(klass).filter_by(
+            organization=signed_claims["organization"],
+            activestate_project_name=signed_claims["project"],
+            actor_id=signed_claims["actor_id"],
+        )
+
+    __lookup_strategies__ = [
+        __lookup_all__,
+    ]
+
+    @property
+    def sub(self) -> str:
+        return f"org:{self.organization}:project:{self.activestate_project_name}"
+
+    @property
+    def builder(self) -> str:
+        return "pypi-publisher"
```

Comment:
Is this necessary if we're verifying it as an invariant?

Reply:
No, this property isn't necessary.

Reply:
I misspoke here, I do need this property.  With my understanding of how the `check_claim_invariant` will work, I need this propery.  The `builder` claim coming from ActiveState will always be `pypi-publisher`.  Am I using `check_claim_invariant` wrong?

Reply:
Ah, you're correct, `check_claim_invariant` is using the value passed to the verifier _and_ the ground truth. This probably isn't actually necessary as the invariant value and the ground truth will always be the same, but changing that is definitely out of scope for this PR (cc @woodruffw for thoughts).